server/tests/test_views.py

```diff
@@ -0,0 +1,77 @@
+# Copyright (C) 2024 Canonical
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Unit tests for Testflinger views
+"""
+
+import mongomock
+from mock import patch
+from src.views import queues_data
+
+
+def test_queues():
+    """
+    Test that the queues view gets the right data from both advertised and
+    unadvertised queues
```

Comment:
I don't fully understand what's the difference between advertised and unadvertised queues. But reading the docstring here suggests that there should be unadvertised queues too. Yet below I see you're only inserting advertised queues

Reply:
So the agents have options that let them "advertise" queues with a name and description. It was easier to keep up with when we didn't have so many agents, but still kind of exists, and otherwise, the server didn't even have any way previously to know about whether a queue exists or not. There's still no concept of "registering" a queue with the server or anything like that, because it's completely opposite of how we built this. The server doesn't *tell* the queues to run a job, the agents declare which queues they know about and ask the server for a job matching one of those tags. But now, the agents also tell the server which ones they are listening to when they start up along with some other data like provisioning type, so that we can inform other systems about this.

...that's a lot of background, but this agent data is where we get the other queues from. It's the second query you see there to retrieve that. Then we make sure not to overwrite any of the ones from advertised queues, because those actually have a description. This was all for a request to make sure that all queues were visible from the UI on this view, and not just the advertised ones.

Reply:
As for the other part of the question, flask doesn't have a native dependency injection system like fastAPI does. There are some 3rd party ones, but I haven't really looked into those yet. I would much rather just redo all this in fastAPI actually - for a host of other reasons, but that's one of them :)

As for the question about mocking vs. using a real mongodb server - that was because these are unit tests and having them run quickly as unit tests from something like pytest means that I can't depend on there being a mongodb server running. The mock just provides the functionality that it would provide if it were there, while also allowing us to inject or read back data from it if needed.

The other question here was about potential for causing problems if we split the views into more files? I'm not sure I follow... In this case, I suppose I could have done something like beautiful soup to try to parse the html and make sure it matched what I wanted, but I think it would be pretty hard to maintain and read those. What I'm much more interested in is whether we are sending the right data to be rendered. I'm not sure how splitting the file would affect that. If we want to split the data gathering part into a different file from the html-rendering part of the view, I think we could do that. I don't see how that would make this harder but I'm happy to take a look if you have more details?

Reply:
Thanks for the very detailed response. If I understand things correctly, now advertised queues and unadvertised queues have become closer as they both are shown on the server and UI. Perhaps we'll some day merge them completely.

I see the benefit of mocking mongo, it really does simplify things as there's no need to deploy a database to run tests.

Regarding the point on splitting the views, I should've been clearer. I was worried about patching the mongo instance as we'd have to patch different places. But this wouldn't be an issue in the future if we move to FastAPI and use dependency injection.